odoo/addons/base/models/ir_ui_menu.py

```diff
@@ -93,15 +93,31 @@ def _visible_menu_ids(self, debug=False):
 
         # process action menus, check whether their action is allowed
         access = self.env['ir.model.access']
-        MODEL_GETTER = {
-            'ir.actions.act_window': lambda action: action.res_model,
-            'ir.actions.report': lambda action: action.model,
-            'ir.actions.server': lambda action: action.model_id.model,
+        MODEL_FIELD = {
+            'ir.actions.act_window': 'res_model',
+            'ir.actions.report': 'model',
+            'ir.actions.server': 'model_name',
         }
+
+        # PREFETCH action models in batch, by category of actions
+        actions_collector = {
+            model: self.env[model]
+            for model in MODEL_FIELD
+        }
+        for menu in action_menus:
+            action = menu.action
+            if action._name in MODEL_FIELD:
+                actions_collector[action._name] += action
+        for model, fname in MODEL_FIELD.items():
+            actions = actions_collector[model]
+            if actions:
+                actions.sudo().mapped(fname)
```

Comment:
Another approach is to determine the "prefetch set" of actions by category
```suggestion
        # determine the ids to prefetch by category
        prefetch_ids = defaultdict(list)
        for action in action_menus.mapped('action'):
            prefetch_ids[action._name].append(action.id)
```
and then in the loop below, add the line
```python
action = action.with_prefetch(prefetch_ids[action._name])
```